### PR TITLE
deps binaries install caching

### PR DESCRIPTION
**Pull request by vtjnash**
this PR provides (relatively) easy caching of deps binaries. The intent is to provide a quick easy way to build a branch or alternative version of a dependency (as a side effect, it also makes switching between versions of most dependencies much faster, without any changes from the user side).

It is on the user to ensure that the build options of the caches are compatible, but for my envisioned usage, I expect that should be fairly straightforward. For example, changing versions of dependencies should be fine, but any other modification (such as USE_BLAS64 or changing the folder structure) is not going to work out quite right.

For example, I could add a build folder for trying out llvm-3.7.0 simply by writing:

``` julia
$ make O=build-llvm37 configure
$ cat - >> build-llvm37/Make.user
build_staging := $(JULIAHOME)/usr-staging
LLVM_VER = 3.7.0
^D
$ make -C build-llvm37
```

this uses the existing usr-staging infrastructure, plus a little extra consistency in the Makefile. the main change is that a the staged source.tgz file is considered up-to-date if it is newer than the corresponding source file(s), ignoring all of the intermediate stages.

as a side benefit, all deps targets now have an `uninstall` and `reinstall` target. (reinstall is more than just uninstall & reinstall, as it will also rebuild the cached install file from the source build files)

if you want to seed a build on a new machine, it should also now be valid to directly copy the `deps/srccache` and `usr-staging` folders, which should help significantly reduce the initial build time on the new system. (I haven't tried it yet, so drop me a line if something goes wonky).

TODO:
- [ ] add note to README / NEWS
- [x] fix cross-deps dependency tracking implementation
- [x] ~~use `tar -U`~~ despite man pages saying this unlinks files, it seems only bsd implements it that way, while gnu also tries to unlink directories

### CODE REVIEW

File: deps/Makefile [hunk 1]
```diff
@@ -879,24 +921,30 @@ OPENLIBM_FLAGS := ARCH="$(ARCH)" CC="$(CC)" FC="$(FC)" AR="$(AR)" OS="$(OS)" USE
 $(OPENLIBM_OBJ_SOURCE): $(BUILDDIR)/$(OPENLIBM_SRC_DIR)/Makefile
 	$(MAKE) -C $(dir $<) $(OPENLIBM_FLAGS) $(MAKE_COMMON)
 	touch -c $@
+
 $(build_shlibdir)/libopenlibm%$(SHLIB_EXT) $(build_libdir)/libopenlibm%a: $(OPENLIBM_OBJ_SOURCE)
```
Comment (vtjnash) [inline deps/Makefile]: delete this old rule

File: deps/Makefile [hunk 2]
```diff
@@ -1713,7 +1804,6 @@ ifeq ($(OS),$(BUILD_OS))
 	$(MAKE) -C $(dir $@) check
 endif
 	echo 1 > $@
-#todo: libunwind tests are known to fail, so they aren't run
 $(LIBUNWIND_TARGET_OBJ): $(LIBUNWIND_TARGET_SOURCE)
 	$(call make-install,libunwind-$(UNWIND_VER),)
```
Comment (vtjnash) [inline deps/Makefile]: fixme: make-install isn't defined anymore

File: deps/Makefile [hunk 3]
```diff
@@ -717,17 +760,19 @@ ifeq ($(OS),$(BUILD_OS))
...
-	$(call make-install,llvm-$(LLVM_VER)/build_$(LLVM_BUILDTYPE),$(LLVM_MFLAGS) PATH="$(llvm_python_workaround):$$PATH")
+$(eval $(call staged-install, \
+	llvm,llvm-$$(LLVM_VER)/build_$$(LLVM_BUILDTYPE), \
+	MAKE_INSTALL,$$(LLVM_MFLAGS) PATH="$$(llvm_python_workaround):$$$$PATH", \
+	$$(LLVM_OBJ_TARGET),))
+stage-llvm: | $(llvm_python_workaround)
 endif # LLVM_USE_CMAKE
-	touch -c $@
-
-reinstall-llvm:
```
Comment (tkelman) [inline deps/Makefile]: is this superseded by something now?

**Comment by tkelman** [comment]
What might this be on osx travis?
```
error: No such file or directory - old dSYM file cannot be overwritten:		old: 'test-double.dSYM/Contents/Resources/DWARF/test-double'
	new:'test-double.dSYM/Contents/Resources/DWARF/test-double.x86_64'.
clang: error: dsymutil command failed with exit code 2 (use -v to see invocation)
94.111667 clang -stdlib=libc++ -mmacosx-version-min=10.7 -m64 -g test-float.c -D__BSD_VISIBLE -I ../include -I../src ../libopenlibm.a -o test-float
make[6]: *** [test-double] Error 2
make[5]: *** [all] Error 2
make[4]: *** [build/openlibm-3ee2a6e6a9fc68cbf2ab00f830b9059215e40b06/libopenlibm.dylib] Error 2
make[3]: *** [/Users/travis/build/JuliaLang/julia/usr-staging/openlibm-3ee2a6e6a9fc68cbf2ab00f830b9059215e40b06.tgz] Error 2
make[2]: *** [/Users/travis/build/JuliaLang/julia/usr-staging/openspecfun-381db9bc865e51de67be9dcaa1610a6f90029c72.tgz] Error 2
```

File: deps/Makefile (hunk 3 above)
Comment (vtjnash) [inline deps/Makefile]: it's auto-generated by staged-install now for all targets

File: deps/Makefile [hunk 4]
```diff
@@ -1006,59 +1058,65 @@ $(BUILDDIR)/Rmath-julia-$(RMATH_JULIA_VER)/Makefile: $(SRCDIR)/srccache/Rmath-ju
...
-install-objconv: $(OBJCONV_TARGET)
+fastcheck-objconv: check-objconv
 
 
 ## OpenBLAS ##
 # LAPACK is built into OpenBLAS by default
 OPENBLAS_GIT_URL := git://github.com/xianyi/OpenBLAS.git
 OPENBLAS_TAR_URL = https://api.github.com/repos/xianyi/OpenBLAS/tarball/$1
-$(eval $(call git-external,openblas,OPENBLAS,Makefile,$(LIBBLASNAME).$(SHLIB_EXT),$(BUILDDIR)))
+$(eval $(call git-external,openblas,OPENBLAS,Makefile,libopenblas$(OPENBLAS_LIBNAMESUFFIX).$(SHLIB_EXT),$(BUILDDIR)))
```
Comment (tkelman) [inline deps/Makefile]: what's the reason for this change?

Comment (vtjnash) [inline deps/Makefile]: the build was getting corrupted if BLAS != openblas

File: deps/Makefile [hunk 5]
```diff
@@ -1516,21 +1594,21 @@ get-fftw: get-fftw-single get-fftw-double
 configure-fftw: configure-fftw-single configure-fftw-double
 compile-fftw: compile-fftw-single compile-fftw-double
 check-fftw: check-fftw-single check-fftw-double
-install-fftw: check-fftw-single check-fftw-double # do these serially, to prevent write conflicts
```
Comment (tkelman) [inline deps/Makefile]: do these not need to be done in serial any more?

File: deps/Makefile [hunk 6]
```diff
@@ -217,22 +223,53 @@ $(foreach dir,$(DIRS),$(eval $(call dir_target,$(dir))))
...
-define staged-install
-	rm -rf $(build_staging)/$1
-	$2
-	mkdir -p $(build_prefix)
-	cp -af $(build_staging)/$1$(build_prefix)/* $(build_prefix)
+# and enables sharing deps compiles, uninstall, and fast reinstall
+MAKE_INSTALL = +$$(MAKE) -C $1 install $$(MAKE_COMMON) $3 DESTDIR="$2"
+define LIBFILE_INSTALL
+	mkdir -p $2/$$(build_libdir)
+	cp $3 $2/$$(build_libdir)
```
Comment (tkelman) [inline deps/Makefile]: no `-f` ?

File: deps/Makefile [hunk 7]
```diff
@@ -1553,24 +1631,30 @@ ifeq ($(OS),$(BUILD_OS))
...
 endif
 	echo 1 > $@
-$(UTF8PROC_OBJ_LIB): $(UTF8PROC_SRC_TARGET)
-	cp -f $< $@
-$(UTF8PROC_OBJ_HEADER): $(UTF8PROC_SRC_TARGET)
-	cp -f $(BUILDDIR)/$(UTF8PROC_SRC_DIR)/utf8proc.h $@
+
+define UTF8PROC_INSTALL
+	mkdir -p $2/$$(build_includedir) $2/$$(build_libdir)
+	cp $1/utf8proc.h $2/$$(build_includedir)
```
Comment (tkelman) [inline deps/Makefile]: no `-f` ?

File: deps/Makefile [hunk 8]
```diff
@@ -1957,7 +2061,7 @@ $(GIT_SOURCE): $(BUILDDIR)/git-$(GIT_VER)/config.status
 $(BUILDDIR)/git-$(GIT_VER)/checked: $(GIT_SOURCE)
 	echo 1 > $@
 $(GIT_TARGET): $(GIT_SOURCE) $(BUILDDIR)/git-$(GIT_VER)/checked
-	$(call make-install,git-$(GIT_VER),NO_INSTALL_HARDLINKS=1 bindir="$(build_libexecdir)")
+	make install -C $(BUILDDIR)/git-$(GIT_VER) $(MAKE_COMMON) NO_INSTALL_HARDLINKS=1 bindir="$(build_libexecdir)"
```
Comment (tkelman) [inline deps/Makefile]: git not worth staging since we'll be getting rid of it soon, or was their build system not cooperating?

File: deps/Makefile (hunk 5 above)
Comment (vtjnash) [inline deps/Makefile]: not sure. we've been installing them in separate folders since usr-staging. would `cp` / `tar x` care?

File: deps/Makefile (hunk 6 above)
Comment (vtjnash) [inline deps/Makefile]: `$2/$$(build_libdir)` is empty

File: deps/Makefile (hunk 8 above)
Comment (vtjnash) [inline deps/Makefile]: not worth the effort to test it (it would have worked)

File: deps/Makefile (hunk 6 above)
Comment (tkelman) [inline deps/Makefile]: could you explain, or point to what guarantees that?

File: deps/Makefile (hunk 5 above)
Comment (tkelman) [inline deps/Makefile]: Not sure, but I think we can merge and find out whether it matters. My only remaining concern is what was happening with dSYM on osx travis.

**Comment by vtjnash** [comment]
> What might this be on osx travis?

i was wondering that too, turns out i messed up a some dependency links in this change. i'll need to devise a new a new way to express cross-deps dependencies, probably by implementing the equivalent of https://github.com/JuliaLang/julia/commit/3aed88c31bdd63b8c30b1f88ac3a8a80d83af200 for the `compile-` and `install-` targets.

File: deps/Makefile (hunk 6 above)
Comment (vtjnash) [inline deps/Makefile]: staged-install guarantees this (as does the `mkdir` on the previous line that just created the directory)

Comment (vtjnash) [inline deps/Makefile]: i can't give then nice variable names, but `$2` is a newly created, pristine DESTDIR

**Comment by tkelman** [comment]
is this completely obsolete?
